include/zephyr/drivers/adc/adc_voltage_divider.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2023 The ChromiumOS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_ADC_VOLTAGE_DIVIDER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_ADC_VOLTAGE_DIVIDER_H_
+
+#include <zephyr/drivers/adc.h>
+
+struct voltage_divider_dt_spec {
+	const struct adc_dt_spec port;
+	int full_ohms;
+	int output_ohms;
+};
+
+/**
+ * @brief Get voltage divider information from devicetree.
+ *
+ * This returns a static initializer for a @p voltage_divider_dt_spec structure
+ * given a devicetree node and a channel index.
+ *
+ * @see VOLTAGE_DIVIDER_DT_SPEC_GET()
+ *
+ * @param node_id Devicetree node identifier.
+ * @param idx index.
+ *
+ * @return Static initializer for an adc_dt_spec structure.
+ */
+#define VOLTAGE_DIVIDER_DT_SPEC_GET_BY_IDX(node_id, idx)                                           \
+	{                                                                                          \
+		.port = ADC_DT_SPEC_GET(node_id),                                                  \
+		.full_ohms = DT_PROP_OR(node_id, full_ohms, 0),                                    \
+		.output_ohms = DT_PROP(node_id, output_ohms),                                      \
+	}
+
+/**
+ * @brief Equivalent to VOLTAGE_DIVIDER_DT_SPEC_GET_BY_IDX(node_id, 0).
+ *
+ * @see VOLTAGE_DIVIDER_DT_SPEC_GET_BY_IDX()
+ *
+ * @param node_id Devicetree node identifier.
+ *
+ * @return Static initializer for an voltage_divider_dt_spec structure.
+ */
+#define VOLTAGE_DIVIDER_DT_SPEC_GET(node_id) VOLTAGE_DIVIDER_DT_SPEC_GET_BY_IDX(node_id, 0)
+
+/**
+ * @brief Calculates the actual voltage from the measured voltage
+ *
+ * @param[in] spec voltage divider specification from Devicetree.
+ * @param[in,out] v_to_v Pointer to the measured voltage on input, and the
+ * corresponding scaled voltage value on output. The stored value is left
+ * unchanged if the binding is used as a resistance based voltage sensor.
+ *
+ * @return 0 on success
+ */
+static inline int divider_scale_voltage_dt(const struct voltage_divider_dt_spec *spec,
+					   int32_t *v_to_v)
+{
+	/* use as resistance based voltage sensor if "full_ohms" is 0 */
```

Comment:
I think this refers to sensors like PTC/NTC resistors, if that's the case maybe we should return an error if full_ohms is not set, and have a second translation function that instead of the full voltage calculates the upper leg resistance. What do you think?

Reply:
Oh, that would require specifying a `full-voltage` though. Maybe just return an error and keep the functionality for a followup.